File: tutorials/b3-running-cxx/run-tmcmc.cpp
#include "model/heat2d.h"
#include "korali.hpp"

int main(int argc, char* argv[])
{
 auto k = korali::Engine();
 auto p = heat2DInit(&argc, &argv);

 k["Problem"]["Type"] = "Evaluation/Bayesian/Inference/Reference";
 k["Problem"]["Likelihood Model"] = "Additive Normal";
 k["Problem"]["Reference Data"] = p.refTemp;
 k["Problem"]["Computational Model"] = &heat2DSolver;

 k["Solver"]["Type"] = "Sampler/TMCMC";
 k["Solver"]["Population Size"] = 200;

 k["Distributions"][0]["Name"] = "Uniform 0";
 k["Distributions"][0]["Type"] = "Univariate/Uniform";
 k["Distributions"][0]["Minimum"] = 10.0;
 k["Distributions"][0]["Maximum"] = 60.0;

 k["Distributions"][1]["Name"] = "Uniform 1";
 k["Distributions"][1]["Type"] = "Univariate/Uniform";
 k["Distributions"][1]["Minimum"] = 0.0;
 k["Distributions"][1]["Maximum"] = 0.5;

 k["Distributions"][2]["Name"] = "Uniform 2";
 k["Distributions"][2]["Type"] = "Univariate/Uniform";
 k["Distributions"][2]["Minimum"] = 0.6;
 k["Distributions"][2]["Maximum"] = 1.0;

 k["Distributions"][3]["Name"] = "Uniform 3";
 k["Distributions"][3]["Type"] = "Univariate/Uniform";
 k["Distributions"][3]["Minimum"] = 0.0;
 k["Distributions"][3]["Maximum"] = 20.0;

 k["Variables"][0]["Name"] = "Intensity";
 k["Variables"][0]["Bayesian Type"] = "Computational";
 k["Variables"][0]["Prior Distribution"] = "Uniform 0";

 k["Variables"][1]["Name"] = "PosX";
 k["Variables"][1]["Bayesian Type"] = "Computational";
 k["Variables"][1]["Prior Distribution"] = "Uniform 1";

 k["Variables"][2]["Name"] = "PosY";
 k["Variables"][2]["Bayesian Type"] = "Computational";
 k["Variables"][2]["Prior Distribution"] = "Uniform 2";

 k["Variables"][3]["Name"] = "Sigma";
 k["Variables"][3]["Bayesian Type"] = "Statistical";
 k["Variables"][3]["Prior Distribution"] = "Uniform 3";

 k.run();

 return 0;
}

File: examples/study.cases/LAMMPS/run-tmcmc.py
#!/usr/bin/env python3
import korali
import sys

sys.path.append("setup/")
from runLAMMPS import *

k = korali.Engine()
e = korali.Experiment()

# Defining Concurrent Jobs
cJobs = 1
if (len(sys.argv) > 1):
  cJobs = int(sys.argv[1])

# Setting up the reference likelihood for the Bayesian Problem
e["Problem"]["Type"] = "Bayesian/Reference"
e["Problem"]["Likelihood Model"] = "Normal"
e["Problem"]["Reference Data"] = getReferenceData()
e["Problem"]["Computational Model"] = runLAMMPS

e["Distributions"][0]["Name"] = "Uniform 0"
e["Distributions"][0]["Type"] = "Univariate/Uniform"
e["Distributions"][0]["Minimum"] = 2.8
e["Distributions"][0]["Maximum"] = 3.0

e["Distributions"][1]["Name"] = "Uniform 1"
e["Distributions"][1]["Type"] = "Univariate/Uniform"
e["Distributions"][1]["Minimum"] = 0.5
e["Distributions"][1]["Maximum"] = 1.25

e["Distributions"][2]["Name"] = "Uniform 2"
e["Distributions"][2]["Type"] = "Univariate/Uniform"
e["Distributions"][2]["Minimum"] = 0.0
e["Distributions"][2]["Maximum"] = 0.1

# Configuring the problem's variables and their prior distributions
e["Variables"][0]["Name"] = "Model-Epsilon"
e["Variables"][0]["Prior Distribution"] = "Uniform 0"

e["Variables"][1]["Name"] = "Model-Sigma"
e["Variables"][1]["Prior Distribution"] = "Uniform 1"

e["Variables"][2]["Name"] = "[Sigma]"
e["Variables"][2]["Prior Distribution"] = "Uniform 2"

# Configuring TMCMC parameters
e["Solver"]["Type"] = "TMCMC"
e["Solver"]["Population Size"] = 32
e["Solver"]["Termination Criteria"]["Max Generations"] = 20

# General Settings
e["File Output"]["Path"] = "_result_lammps_tmcmc_" + str(cJobs).zfill(
    4) + "_workers"
e["Console Output"]["Verbosity"] = "Detailed"

# Selecting external conduit
k["Conduit"]["Type"] = "Concurrent"
k["Conduit"]["Concurrent Jobs"] = cJobs

k.run(e)
